utils/version.js

```diff
@@ -16,28 +16,59 @@ module.exports.mergeVersions = (name, versions) => {
     return versions[0];
   }
   const gitBranches = versions.filter((d) => d.includes('github') || d.includes('gitlab'));
+  const rest = versions.filter((v) => !gitBranches.includes(v));
 
   if (gitBranches.length) {
-    const haveCommonVersions = (branches) => {
-      versions = new Set(branches.map((branch) => branch.split('#')[1]));
-      return versions.size === 1;
-    };
-    if (haveCommonVersions(gitBranches)) {
-      return gitBranches[0];
-    }
-
-    throw new Error(chalk.red(`Versions ${chalk.white(gitBranches.join(', '))} point to different branches, which can lead to workspace errors.\nPlease use the same branch in all versions.`));
+    return mergedGithubVersions(name, gitBranches, rest);
   }
 
   try {
-    return intersect(...versions).toString();
+    return intersect(...rest).toString();
   } catch (e) {
     // map to base version, sort descending take first
-    const max = findMaxVersion(versions);
+    const max = findMaxVersion(rest);
     console.warn(`cannot find common intersecting version for ${name} = ${versions.join(', ')}, taking max "${max}" for now`);
     return max.toString();
   }
-};
+}
+
+module.exports.mergeVersions = mergeVersions;
+
+
+function compareGitWithNpm(gitVersion, npmVersion) {
+  const areEqual = (v1, v2) => semver.eq(removeRangeTag(v1), removeRangeTag(v2));
+  return (!npmVersion || areEqual(gitVersion, npmVersion));
+}
+
+/**
+ * Finds the intersection of an array of versions pointing to github or gitlab
```

Comment:
Can you maybe outline the algorithm step by step? This might also make the implementation clearer.